### PR TITLE
Port changes from my thesis branch

### PR DESCRIPTION
**Pull request by XianzheMa**

### CODE REVIEW

File: modyn/common/ftp/ftp_utils.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ def delete_file(hostname: str, port: int, user: str, password: str, remote_file_
     Returns:
     """
     ftp = FTP()
-    ftp.connect(hostname, port, timeout=3)
+    ftp.connect(hostname, port, timeout=5 * 60)
```
Comment (XianzheMa) [inline modyn/common/ftp/ftp_utils.py:106]: otherwise for criteo it times out

File: modyn/evaluator/internal/metrics/f1_score.py [hunk 2]
```diff
@@ -70,4 +70,6 @@ def get_evaluation_result(self) -> float:
         return float(np.average(f1_scores, weights=total_labels_per_class / total_samples))
 
     def get_name(self) -> str:
+        if self.config.average == "binary":
+            return f"F1-{self.config.average}-{self.config.pos_label}"
```
Comment (XianzheMa) [inline modyn/evaluator/internal/metrics/f1_score.py:74]: for f1 binary we can specify either 0 or 1 as the positive label. For those two metrics they should not have the same name, should be distinguished by pos label

File: modyn/evaluator/internal/pytorch_evaluator.py [hunk 3]
```diff
@@ -64,7 +64,6 @@ def _prepare_dataloader(
             dataset,
             batch_size=evaluation_info.batch_size,
             num_workers=evaluation_info.num_dataloaders,
-            timeout=60 if evaluation_info.num_dataloaders > 0 else 0,
```
Comment (XianzheMa) [inline modyn/evaluator/internal/pytorch_evaluator.py:67]: otherwise for criteo it times out

File: modyn/evaluator/internal/metrics/f1_score.py (hunk 2 above)
Comment (MaxiBoether) [reply modyn/evaluator/internal/metrics/f1_score.py:74]: Can you add this as a comment please :)?

File: modyn/evaluator/internal/pytorch_evaluator.py (hunk 3 above)
Comment (MaxiBoether) [reply modyn/evaluator/internal/pytorch_evaluator.py:67]: we should not merge this. I don't understand why or how it would time out, need to investigate myself. We should not wait for over 60 seconds to receive data, even in the criteo case. Have not encountered this in my benchmarks so far!

File: modyn/evaluator/internal/metrics/roc_auc.py [hunk 4]
```diff
@@ -33,7 +33,10 @@ def _dataset_evaluated_callback(self, y_true: torch.Tensor, y_pred: torch.Tensor
         desc_score_indices = torch.argsort(y_pred, descending=True)
         y_score = y_pred[desc_score_indices]
         y_true = y_true[desc_score_indices]
-        distinct_value_indices = torch.nonzero(y_score[1:] - y_score[:-1], as_tuple=False).squeeze()
+        # we only need to squeeze the second dimension;
+        # otherwise if there is only one non-zero element in (y_score[1:] - y_score[:-1]),
+        # after squeezing it will become a scalar, which will cause an error in torch.cat
+        distinct_value_indices = torch.nonzero(y_score[1:] - y_score[:-1], as_tuple=False).squeeze(dim=1)
```
Comment (MaxiBoether) [inline modyn/evaluator/internal/metrics/roc_auc.py:39]: I don't understand this. This has been working before? Can you show a test that fails with the old implementation and works with the new one? Because this metric implementation has been used like this for quite some time

File: modyn/trainer_server/internal/dataset/data_utils.py [hunk 5]
```diff
@@ -70,7 +70,6 @@ def prepare_dataloaders(
         batch_size=batch_size,
         num_workers=num_dataloaders,
         drop_last=drop_last,
-        timeout=60 if num_dataloaders > 0 else 0,
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/dataset/data_utils.py:73]: see my other comment

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thank you for these changes, please see my individual comments. Is this everything from your thesis branch? I also remember discussions on undefined/nan values in the model output. Did you change anything for that?

File: modyn/evaluator/internal/metrics/f1_score.py (hunk 2 above)
Comment (XianzheMa) [reply modyn/evaluator/internal/metrics/f1_score.py:74]: will do

**Comment by XianzheMa** [comment]
> Thank you for these changes, please see my individual comments. Is this everything from your thesis branch? I also remember discussions on undefined/nan values in the model output. Did you change anything for that?

No, changes in the PR are just miscellaneous stuff. I didn't come up with a way to work around the nans but just used full time training.

There will be another PR on returning the relative positions than the sample ids for selection.

File: modyn/evaluator/internal/metrics/roc_auc.py (hunk 4 above)
Comment (XianzheMa) [reply modyn/evaluator/internal/metrics/roc_auc.py:39]: the case when it does not work is very rare but does exist. Will come up with a test.

File: modyn/tests/evaluator/internal/metrics/test_roc_auc.py [hunk 6]
```diff
@@ -39,6 +39,12 @@ def test_roc_auc_metric():
     assert get_evaluation_result(y_true, y_score) == pytest.approx(2 / 3)
 
 
+def test_roc_auc_with_two_entries():
+    y_true = np.array([0, 1])
+    y_score = np.array([0.1, 0.6])
+    get_evaluation_result(y_true, y_score)
```
Comment (XianzheMa) [inline modyn/tests/evaluator/internal/metrics/test_roc_auc.py:45]: without the change, this line will raise an exception

**Review by MaxiBoether — APPROVED** [review]
Thank you!

File: modyn/tests/evaluator/internal/metrics/test_roc_auc.py (hunk 6 above)
Comment (MaxiBoether) [reply modyn/tests/evaluator/internal/metrics/test_roc_auc.py:47]: thank you, now I understand the issue!

File: modyn/evaluator/internal/metrics/roc_auc.py [hunk 7]
```diff
@@ -33,7 +33,10 @@ def _dataset_evaluated_callback(self, y_true: torch.Tensor, y_pred: torch.Tensor
         desc_score_indices = torch.argsort(y_pred, descending=True)
         y_score = y_pred[desc_score_indices]
```
Comment (robinholzi) [inline modyn/evaluator/internal/metrics/roc_auc.py:34]: @XianzheMa  Is this related?
image

Comment (XianzheMa) [reply modyn/evaluator/internal/metrics/roc_auc.py:34]: hmmm I don't think so. What dataset/model did you use in this experiment?

Comment (robinholzi) [reply modyn/evaluator/internal/metrics/roc_auc.py:34]: I think we solved it now. Seems to be related with an invalid choice of evaluation transformer function
